### PR TITLE
OBPIH-5043 Validate for duplicate associations

### PR DESCRIPTION
**Pull request by drodzewicz**

### CODE REVIEW

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy [hunk 1]
```diff
@@ -28,8 +29,32 @@ class ProductAssociationDataService {
...
             productAssociationInstance.associatedProduct = Product.findByProductCode(params['associatedProduct.productCode'])
             params['product.id'] =  productAssociationInstance.product?.id
+            params['product.name'] =  productAssociationInstance.product?.name
             params['associatedProduct.id'] =  productAssociationInstance.associatedProduct?.id
+            params['associatedProduct.name'] =  productAssociationInstance.associatedProduct?.name
 
+            params.hasMutualAssociation = Boolean.valueOf(params.hasMutualAssociation as String) || (params.hasMutualAssociation as String)?.equalsIgnoreCase("yes")
+
+            // check for duplicate association entries in the file
+            def indexOfDuplicate = listOfValidatedProductAssociationParams.findIndexOf{it ->
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:39]: Is the `findIndexOf` done on an empty list, or am I missing something here?

**Review by kchelstowski — CHANGES_REQUESTED** [review]

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy [hunk 2]
```diff
@@ -28,8 +29,32 @@ class ProductAssociationDataService {
...
+                if (params.hasMutualAssociation || it.hasMutualAssociation) {
+                    return  (
+                            params['product.id'] == it['product.id'] &&
+                            params['associatedProduct.id'] == it['associatedProduct.id'] &&
+                            params['code'] == params['code']
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:44]: Seems like this (`params['code'] == params['code']`) will be always true

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy [hunk 3]
```diff
@@ -65,16 +90,42 @@ class ProductAssociationDataService {
...
+                            eq("associatedProduct", productAssociationInstance.product)
+                            eq("code", productAssociationInstance.code)
+                        }
+                    }
+                }
+                if (existingAssociations) {
+                    if (existingAssociations.any{ it.mutualAssociation }) {
+                        command.errors.reject("Row ${index + 1}: Same two-way association already exists")
+                    } else {
+                        command.errors.reject("Row ${index + 1}: One-way of the association already exists")
```
[the remark this answers is not in the document]
Comment (kchelstowski) [reply grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:113]: or nvm, now I realized the context - you wouldn't want to display this error from else if there is not any association

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy [hunk 4]
```diff
@@ -28,8 +29,32 @@ class ProductAssociationDataService {
...
+            def indexOfDuplicate = listOfValidatedProductAssociationParams.findIndexOf{it ->
+                if (params.hasMutualAssociation || it.hasMutualAssociation) {
+                    return  (
+                            params['product.id'] == it['product.id'] &&
+                            params['associatedProduct.id'] == it['associatedProduct.id'] &&
+                            params['code'] == params['code']
+                        ) || (
+                            params['product.id'] == it['associatedProduct.id'] &&
+                            params['associatedProduct.id'] == it['product.id'] &&
+                            params['code'] == params['code']
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:48]: Same here

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy [hunk 5]
```diff
@@ -28,8 +29,32 @@ class ProductAssociationDataService {
...
+            params['product.name'] =  productAssociationInstance.product?.name
             params['associatedProduct.id'] =  productAssociationInstance.associatedProduct?.id
+            params['associatedProduct.name'] =  productAssociationInstance.associatedProduct?.name
 
+            params.hasMutualAssociation = Boolean.valueOf(params.hasMutualAssociation as String) || (params.hasMutualAssociation as String)?.equalsIgnoreCase("yes")
+
+            // check for duplicate association entries in the file
+            def indexOfDuplicate = listOfValidatedProductAssociationParams.findIndexOf{it ->
+                if (params.hasMutualAssociation || it.hasMutualAssociation) {
+                    return  (
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:41]: This entire `or` looks weird are you sure it is proper?

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy [hunk 6]
```diff
@@ -28,8 +29,32 @@ class ProductAssociationDataService {
...
+                            params['code'] == params['code']
+                        ) || (
+                            params['product.id'] == it['associatedProduct.id'] &&
+                            params['associatedProduct.id'] == it['product.id'] &&
+                            params['code'] == params['code']
+                        )
+                }
+                return params['product.id'] == it['product.id'] &&
+                       params['associatedProduct.id'] == it['associatedProduct.id'] &&
+                       params['code'] == params['code']
```
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:53]: Same as above

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy (hunk 1 above)
Comment (awalkowiak) [inline grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:39]: Add a space before and after the `{`

**Review by awalkowiak — CHANGES_REQUESTED** [review]

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy (hunk 2 above)
Comment (drodzewicz) [reply grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:44]: yep this is a typo

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy (hunk 1 above)
Comment (drodzewicz) [reply grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:39]: on line 128 I add visited/checked association to this list and later check for duplicates
```
listOfValidatedProductAssociationParams.add(params)
```

File: grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy (hunk 5 above)
Comment (drodzewicz) [reply grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy:41]: It works properly but I agree it looks ugly.
the problem I had here was that I had to detect not only duplicates but also a case where we have two-way and one-way entries of the same association
```
Association_1(product: A, associatedProduct: B, type: Type1, hasMutualAssociation: false )
Association_2(product: A, associatedProduct: B, type: Type1, hasMutualAssociation: true )
```
```
Association_1(product: A, associatedProduct: B, type: Type1, hasMutualAssociation: false )
Association_2(product: B, associatedProduct: A, type: Type1, hasMutualAssociation: true )
```

**Review by jmiranda — COMMENTED** [review]
I am not in a good mindset to really dig into this. My one comment would be ... add very descriptive comments about what we're doing here. Going so far as to describe the anatomy or examples the expected input so we can see what and why we're doing what we're doing. 

One of your comments below does a good job explaining the what or why. So that needs to go into the code as a comment. I would go even further and document all of the scenarios that might come up.

**Pull request merged**